docs/design/pr-builder.rst

```diff
@@ -0,0 +1,146 @@
+Design of Pull Request Builder
+==============================
+
+Background
+----------
+
+This will focus on automatically building documentation for Pull Requests on Read the Docs projects.
+This is one of the most requested feature of Read the Docs.
+This document will serve as a design document for discussing how to implement this features.
+
+Scope
+-----
+
+- Receiving ``pull_request`` webhook event from Github
+- Fetching data from pull requests.
+- Making Pull Requests work like temporary ``Version``
+- Setting PR version Life time
+- Excluding PR Versions from Elasticsearch Indexing
+- Creating PR Versions when a pull request is opened and Triggering a build
+- Triggering Builds on new PR commits
+- Status reporting to Github
+- Storing PR Version build Data
+- Deleting PR version and the build data
+- Excluding PR Versions from Search Engines
+- Serving PR Docs
+- Update the Footer API
+
+Fetching Data from Pull Requests
+--------------------------------
+
+We already get Pull request events from Github webhooks.
+We can utilize that to fetch data from pull requests.
+when a ``pull_request`` event is triggered we can fetch the data of that pull request.
+We can fetch the pull request by doing something similar to travis-ci.
+ie: ``git fetch origin +refs/pull/<pr_number>/merge:``
+
+Modeling Pull Requests as a Type of Version
+-------------------------------------------
+
+Pull requests can be Treated as a Type of Temporary ``Version``.
+We might consider adding a Boolean Field or ``VERSION_TYPES`` to the ``Version`` model.
+
+- If we go with Boolean Field we can add something like ``is_pull_request`` Field.
+- If we go with ``VERSION_TYPES`` we can add something like ``pull_request`` alongside Tag and Branch.
+
+We also have to update the current ``Version`` model ``QuerySet`` to exclude the PR versions.
+We have to add ``QuerySet`` for PR versions also.
+
+Excluding PR Versions from Elasticsearch Indexing
+-------------------------------------------------
+
+We should exclude to PR Versions from being Indexed to Elasticsearch.
```

Comment:
I think this is the easiest way and it's fine to me.

Although, I was wondering if we want to have it indexed but excluded from the global search results. I mean, if you perform a search over a PR version documentation you probably expect that to work and return the proper results, right?

On the other hand, if you perform a search over the whole project for a particular term, you don't want to expose results from the PR versions.

What I'm saying it's probably too complicated and should not be a focus for the first implementation, though.

Reply:
I agree. we can implement this after the initial implementation is shipped.